### PR TITLE
Make s3 gateway DNS optional

### PR DESCRIPTION
**Pull request by ozkatz**
This PR (possibly) supersedes #2351, #2352 and #2312.

If merged, most users will not need to configure any DNS record to use lakeFS, except for 2 (relatively advanced?) cases:

1. They're working with a system that doesn't support [path-based addressing for S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html). 
2. They wish to use public-access repositories that don't require authentication *Note this is not currently supported by lakeFS*

For everyone else, it makes it possible to spin up a lakeFS instance, and use whatever address the system provides for it, for both S3 and API use:

```
$ aws s3 --endpoint-url="http://localhost:8000" ls
```

will work, but also 

```
$ lakectl fs ls lakefs://my-local-repo/main/...
```

Without any special configuration.

### How does it work?

It works by changing the way we decide whether a request goes to the UI/API handler, or the S3 Gateway handler. It will now resolve it using the following order:

1. If the request is made with a `Host` header that matches `gateways.s3.domain_name`, route to S3 gateway
2. If the request is made with a `Host` header that is a sub-domain of `gateways.s3.domain_name`, route to S3 gateway
3. **If the request contains an AWS Sigv2 or Sigv4 signature, route to S3 gateway**
4. Otherwise, route to the UI/API handler

The addition is of course # 3 above. Both sigv2 and sigv4 have very specific query params or HTTP header values, not used anywhere else. This makes them good candidates for this type of inference: They will ALWAYS be present in S3 authenticated requests, and will likely NEVER be present in API, SDK or UI requests.

For users who do wish to use unauthenticated access (if we do end up supporting it) or that cannot use path-based addressing, they can set up the required DNS records the same way it's done today. 

### Why do it?

Because the user journey is currently overly partly because of the DNS settings. 

This makes running lakeFS locally (using Docker or otherwise) pretty complex since you need to understand its internal routing of S3 requests prior to using the project.

It also complicates setting up a POC environment (and possibly even production deployments) for cases where the virtual-host addressing is not required. 

### Testing

The code modifications are relatively easy to test and are covered by both unit and system tests.

The harder part is testing all the integration documentation. For that I'd love some help.

Tested on:
- [x] EMR
- [x] Databricks
- [x] Boto
- [x] MinIO client
- [x] Kafka connect

### CODE REVIEW

**Comment by johnnyaug** [comment]
I'd like to suggest rolling this feature silently, to allow us to test it thoroughly.
We can separate the docs to a second PR and merge it in a couple of weeks.

**Comment by ozkatz** [comment]
> I'd like to suggest rolling this feature silently, to allow us to test it thoroughly.
> We can separate the docs to a second PR and merge it in a couple of weeks.

That's possible! But I'm not sure what the benefit is, over testing this throughly before merging.. a silent release would be mostly testing the change is backwards compatible with current docs, which we can think of other ways to validate

File: cmd/lakefs/cmd/run.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/treeverse/lakefs/pkg/gateway/sig"
+
```
Comment (nopcoder) [inline cmd/lakefs/cmd/run.go:16]: remove white space

File: cmd/lakefs/cmd/run.go [hunk 2]
```diff
@@ -174,12 +176,18 @@ var runCmd = &cobra.Command{
...
 			Addr: cfg.GetListenAddress(),
-			Handler: httputil.HostMux(
-				httputil.HostHandler(apiHandler).Default(), // api as default handler
-				httputil.HostHandler(s3gatewayHandler, // s3 gateway for its bare domain and sub-domains of that
-					httputil.Exact(cfg.GetS3GatewayDomainNames()),
-					httputil.SubdomainsOf(cfg.GetS3GatewayDomainNames())),
-			),
+			Handler: http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+				// If the request has the S3 GW domain (exact or subdomain) - or carries an AWS sig, serve S3GW
+				if httputil.HostMatches(request, cfg.GetS3GatewayDomainNames()) ||
```
Comment (nopcoder) [inline cmd/lakefs/cmd/run.go:181]: Looks like you switched from using the generic solution called HostMux - can we throw some code?

File: pkg/httputil/server_test.go [hunk 3]
```diff
@@ -33,15 +35,19 @@ func TestSubdomainsOf(t *testing.T) {
...
+			got := HostSubdomainOf(r, tt.args.v)
+			if got != tt.want {
+				t.Errorf("HostSubdomainOf() '%s' test with '%s' got = %t, want = %t", tt.args.v, tt.host, got, tt.want)
 			}
 		})
 	}
 }
 
-func TestExact(t *testing.T) {
+func TestHostMatches(t *testing.T) {
```
Comment (nopcoder) [inline pkg/httputil/server_test.go:50]: Need a test for empty list

File: docs/deploy/includes/aws-docker-run.md [hunk 4]
```diff
@@ -1,10 +1,9 @@
 ```sh
-docker run \
+`docker run \
```
Comment (nopcoder) [inline docs/deploy/includes/aws-docker-run.md:2]: why the extra ` - already inside a code block

File: docs/deploy/includes/prerequisites.md [hunk 5]
```diff
@@ -1,10 +1,13 @@
 ## Prerequisites
 
 {: .no_toc }
-A production-suitable lakeFS installation will require three DNS records **pointing at your lakeFS server**.
+
+Users that require S3 access using [virtual host addressing](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) will require three DNS records **pointing at your lakeFS server**.
 A good convention for those will be, assuming you already own the domain `example.com`:
 * `lakefs.example.com`
-* `s3.lakefs.example.com` - **this is the S3 Gateway Domain**
+* `s3.lakefs.example.com` - **this is the S3 Gateway explicit Domain**
```
Comment (nopcoder) [inline docs/deploy/includes/prerequisites.md:8]:
```suggestion
* `s3.lakefs.example.com` - **this is the S3 Gateway explicit domain**
```

File: docs/integrations/databricks.md [hunk 6]
```diff
@@ -28,7 +28,8 @@ For Databricks to work with lakeFS, set the S3 Hadoop configuration to the lakeF
 ```
 spark.hadoop.fs.s3a.bucket.<repo-name>.access.key AKIAIOSFODNN7EXAMPLE
 spark.hadoop.fs.s3a.bucket.<repo-name>.secret.key wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
-spark.hadoop.fs.s3a.bucket.<repo-name>.endpoint https://s3.lakefs.example.com
+spark.hadoop.fs.s3a.bucket.<repo-name>.endpoint https://lakefs.example.com
+spark.hadoop.fs.s3a.path.style.access true
```
Comment (nopcoder) [inline docs/integrations/databricks.md:32]: this one doesn't have a bucket-level setting?

File: docs/integrations/dremio.md [hunk 7]
```diff
@@ -20,10 +20,11 @@ Suppose you already have both lakeFS and Dremio deployed, and want to utilize Dr
...
-1. Under `Advanced Options` > `Connection Properties`, add `fs.s3a.endpoint` and set lakeFS S3 endpoint to the value.
-1. Under the `General` tab, specify the `access_key_id` and `secret_access_key` provided by lakeFS server.
-1. Click `Save`, and now you should be able to browse lakeFS repositories on Dremio.
+2. Under `File Stores`, choose `Amazon S3`.
+3. Under `Advanced Options`, check `Enable compatibility mode (experimental)`.
+4. Under `Advanced Options` > `Connection Properties`, add `fs.s3a.path.style.access` and set the value to true.
+5. Under `Advanced Options` > `Connection Properties`, add `fs.s3a.endpoint` and set lakeFS S3 endpoint to the value.
+6. 1. Under `Advanced Options` > `Connection Properties`, add `fs.s3a.path.style.access` and set to `true`
+7. Under the `General` tab, specify the `access_key_id` and `secret_access_key` provided by lakeFS server.
+8. Click `Save`, and now you should be able to browse lakeFS repositories on Dremio.
```
Comment (nopcoder) [inline docs/integrations/dremio.md:29]: 1. Markdown numbering - use: 1.
1. Item number 6 - there is an extra numbering

File: docs/setup/virtual-host-addressing.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+Typically, if the lakeFS installation is served under `lakefs.example.com`, a good choice would be `s3.lakefs.example.com`.
+
+This could be done using either an environment variable:
+
+```shell
+LAKEFS_GATEWAYS_S3_DOMAIN_NAME="s3.lakefs.example.com"
+```
+
+Or by adding the `gateways.s3.domain_name` setting to the `lakefs.yaml` config file:
```
Comment (nopcoder) [inline docs/setup/virtual-host-addressing.md:40]:
```suggestion
Or by adding the `gateways.s3.domain_name` setting to the lakeFS's `config.yaml` config file:
```

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: cmd/lakefs/cmd/run.go (hunk 2 above)
Comment (ozkatz) [reply cmd/lakefs/cmd/run.go:181]: Already thrown :)

File: docs/integrations/databricks.md (hunk 6 above)
Comment (ozkatz) [reply docs/integrations/databricks.md:32]: it might but if you're requiring this behavior from the SDK might as well make it consistent?

File: docs/setup/virtual-host-addressing.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+has_children: false
+---
+
+# Configuring lakeFS to use S3 Virtual-Host addressing
+
+## Understanding virtual-host addressing
+
+Some systems require S3 endpoints (such as lakeFS' [S3 Gateway](../understand/architecture.md#s3-gateway)) to support [virtual-host style addressing](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html){: target="_blank" }.
+
+lakeFS supports this, but requires some configuration in order to extract the bucket name (used as the repository ID) from the host address.
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:15]:
```suggestion
lakeFS supports this, but requires some configuration in order to extract the bucket name (used as the lakeFS repository ID) from the host address.
```

File: docs/setup/virtual-host-addressing.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+For example:
+
+```text
+GET http://foo.example.com/some/location
+```
+
+In this case, there's no way for lakeFS to distinguish this is a virtual-host request where the endpoint url is `example.com`, the bucket name is `foo` and the path is `/some/location`,
+or this is a path-based request where the endpoint is `foo.example.com`, the bucket name is `some` and the path is `location`.
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:24]:
```suggestion
In this case, there's no way for lakeFS to determine whether this is a virtual-host request where the endpoint url is `example.com`, the bucket name is `foo` and the path is `/some/location`,
or a path-based request where the endpoint is `foo.example.com`, the bucket name is `some` and the path is `location`.
```

File: docs/setup/virtual-host-addressing.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+This requires an extra step: Defining an explicit set of DNS record for lakeFS S3 gateway.
+
+## Adding an explicit S3 domain name to the S3 Gateway configuration
+
+The first step would be to tell the lakeFS installation which hostnames are used as an S3. This should be a different DNS record from the one used for e.g. the UI or API.
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:30]:
```suggestion
The first step would be to tell the lakeFS installation which hostnames are used for the S3 gateway. This should be a different DNS record from the one used for the UI or API.
```

File: docs/setup/virtual-host-addressing.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+...
+
+# This section defines an explict S3 gateway address that supports virtual-host addressing
+gateways:
+  s3:
+    domain_name: s3.lakefs.example.com
+```
+
+for more information on how to configure lakeFS, check out the [configuration reference](../reference/configuration.md)
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:55]:
```suggestion
For more information on how to configure lakeFS, check out the [configuration reference](../reference/configuration.md)
```

File: docs/setup/virtual-host-addressing.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+{: .note }
+
+## Setting up the appropriate DNS records
+
+Once our lakeFS installation is configured with an explicit S3 gateway endpoint address, we need to define 2 DNS records and have them point at our lakeFS installation.
+This requires 2 CNAME records:
+
+1. `s3.lakefs.example.com` - CNAME to `lakefs.example.com`. This would be used as the S3 endpoint when configuring clients and will serve as our bare domain.
+1. `*.s3.lakefs.example.com` - Also a CNAME to `lakefs.example.com`. This will resolve virtual-host requests such as `my-bucket.s3.lakefs.example.com` that lakeFS would now know how to parse.
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:64]:
```suggestion
1. `*.s3.lakefs.example.com` - Also a CNAME to `lakefs.example.com`. This will resolve virtual-host requests such as `example-repo.s3.lakefs.example.com` that lakeFS would now know how to parse.
```

File: docs/setup/virtual-host-addressing.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+
+<div class="note">
```
Comment (johnnyaug) [inline docs/setup/virtual-host-addressing.md:67]: Any reason why this is not markdown?

File: docs/deploy/includes/prerequisites.md [hunk 15]
```diff
@@ -1,10 +1,13 @@
 ## Prerequisites
 
 {: .no_toc }
-A production-suitable lakeFS installation will require three DNS records **pointing at your lakeFS server**.
+
+Users that require S3 access using [virtual host addressing](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) will require three DNS records **pointing at your lakeFS server**.
```
Comment (johnnyaug) [inline docs/deploy/includes/prerequisites.md:5]: Perhaps link to the new doc page and not AWS?

Or maybe even remove this prerequisites section and leave just a link to the new doc page

**Review by johnnyaug — COMMENTED** [review]
This looks very good. I'm not approving because I think this change is too big and needs to be broken down somehow. IMO it will be very hard for one person to test this manually. Just to clarify, I'm not concerned about backward compatibility, but about some obscure operations on some S3 clients not playing nice with this.

What we can do is release only the server changes without the docs or the quickstart docker-compose changes. This way we can test it on our day-to-day:
1. Remove the gateway domain from our dev configuration files.
2. Change our internal terraform to work w/o the gateway domain.
3. Test this change with design partners we work with 

Later we can release the docs, and only after change the quickstart docker compose as well.

I'm also open to other suggestion for how to take this in small steps.

**Comment by ozkatz** [comment]
As agreed, this will be the testing plan:

Databricks - @arielshaqed 
EMR - @itaiad200 
Trino - @guy-har 
Align Terraform and setup lakeFS - @johnnyaug 
minio client, boto, Kafka connect - @ozkatz

File: docs/setup/virtual-host-addressing.md (hunk 14 above)
Comment (ozkatz) [reply docs/setup/virtual-host-addressing.md:67]: Yes - I'm not good enough with Markdown to add a .note class that wraps two individual paragraphs. 🤦‍♂️

**Review by talSofer — COMMENTED** [review]
Need to add this change to CHANGELOG :)

**Comment by ozkatz** [comment]
> Need to add this change to CHANGELOG :)

@talSofer thanks, added

File: docs/deploy/includes/aws-docker-run.md [hunk 16]
```diff
@@ -5,6 +5,5 @@ docker run \
   -e LAKEFS_DATABASE_CONNECTION_STRING="[DATABASE_CONNECTION_STRING]" \
   -e LAKEFS_AUTH_ENCRYPT_SECRET_KEY="[ENCRYPTION_SECRET_KEY]" \
   -e LAKEFS_BLOCKSTORE_TYPE="s3" \
-  -e LAKEFS_GATEWAYS_S3_DOMAIN_NAME="[S3_GATEWAY_DOMAIN]" \
-  treeverse/lakefs:latest run
+  treeverse/lakefs:latest run`
```
Comment (adunuthulan) [inline docs/deploy/includes/aws-docker-run.md:8]:
```suggestion
  treeverse/lakefs:latest run
```

**Comment by ozkatz** [comment]
good catch @adunuthulan!

File: docs/integrations/databricks.md (hunk 6 above)
Comment (nopcoder) [reply docs/integrations/databricks.md:32]: thought the above example is how to make it specific - but I'll agree as long it is the preferred way anyway.

**Review by nopcoder — APPROVED** [review]

**Comment by ozkatz** [comment]
All tests we defined pass! Merging. Thanks @arielshaqed @itaiad200 @guy-har @johnnyaug @nopcoder @adunuthulan for helping with reviews and tests!
